mbtiles/src/bin/mbtiles.rs

```diff
@@ -332,6 +385,216 @@ async fn meta_set_value(file: &Path, key: &str, value: Option<&str>) -> MbtResul
+                    Some("pbf" | "mvt") => Some("pbf".to_string()),
+                    Some("jpg" | "jpeg") => Some("jpg".to_string()),
+                    Some("webp") => Some("webp".to_string()),
+                    Some("png") => Some("png".to_string()),
+                    _ => {
+                        anyhow::bail!("Unsupported file extension: {}", entry.path().display());
+                    }
+                };
+
+                if format == Some("pbf".to_string()) {
+                    compress = true;
+                }
+            }
+
+            let data = std::fs::read(entry.path())?;
+
+            let encoded = if compress {
+                let mut encoder = GzEncoder::new(Vec::new(), Compression::default());
+                encoder.write_all(&data)?;
+                encoder.finish()?
+            } else {
+                data
+            };
+
+            // Convert from TMS to XYZ if necessary
+            let y = match scheme {
+                TileScheme::Xyz => mbtiles::invert_y_value(z, y),
+                TileScheme::Tms => y,
+            };
+
+            mbt.insert_tiles(
+                &mut conn,
+                MbtType::Flat,
+                CopyDuplicateMode::Abort,
+                &[(z, x, y, encoded)],
+            )
+            .await?;
+        }
+    }
+
+    if let Some(format) = format {
+        mbt.set_metadata_value(&mut conn, "format", format).await?;
+    }
+
+    // TODO: set minzoom, maxzoom, and bbox?
+    // either compute them, or possibly read them from {input_directory}/metadata.json
+
+    Ok(())
+}
+
+async fn unpack(
+    input_file: &Path,
+    output_directory: &Path,
+    scheme: TileScheme,
+) -> anyhow::Result<()> {
+    if !input_file.exists() {
+        anyhow::bail!("Input file does not exist: {}", input_file.display());
+    }
+
+    let mbt = Mbtiles::new(input_file)?;
+    let mut conn = mbt.open_readonly().await?;
+
+    // Get the format from metadata to determine file extension and compression
+    let format = mbt.get_metadata_value(&mut conn, "format").await?;
+    let (extension, decompress) = match format.as_deref() {
+        Some("pbf") => ("mvt", true),
+        Some("jpg") => ("jpg", false),
+        Some("png") => ("png", false),
+        Some("webp") => ("webp", false),
+        Some(unknown) => {
+            anyhow::bail!("Unknown format in MBTiles metadata: {}", unknown);
+        }
+        None => anyhow::bail!("No format specified in MBTiles metadata"),
+    };
+
+    // Create output directory if it doesn't exist
+    std::fs::create_dir_all(output_directory)?;
+
+    // Query all tiles from the database
+    let mut tiles = sqlx::query!("SELECT zoom_level, tile_column, tile_row, tile_data FROM tiles ORDER BY zoom_level, tile_column, tile_row")
+        .fetch(&mut conn);
+
+    while let Some(tile) = tiles.try_next().await? {
+        let Some(z) = tile.zoom_level else {
+            log::warn!("Skipping tile with missing zoom level");
```

Comment:
while this would be mega-rare, you might still have some sort of a flag to avoid flooding the output with duplicate warnings